File: Tests/CaptureTests.cs
using System;
using Stringier.Patterns;
using static Stringier.Patterns.Pattern;
using Xunit;

namespace Tests {
	public class CaptureTests {
		[Fact]
		public void Constructor() => 'a'.Or('b').Or('c').Capture(out Capture capture);

		[Fact]
		public void Consume() {
			Capture capture;
			Pattern pattern = 'a'.Or('b').Or('c').Capture(out capture);
			ResultAssert.Captures("a", pattern.Consume("a"));
			CaptureAssert.Captures("a", capture);

			pattern = 'a'.Or('b').Or('c').Capture(out capture).Then('!');
			ResultAssert.Fails(pattern.Consume("a")); 
			ResultAssert.Captures("a!", pattern.Consume("a!"));
			CaptureAssert.Captures("a", capture);

			pattern = 'a'.Or('b').Or('c').Capture(out capture).Then(',').Then(capture);
			ResultAssert.Captures("b,b", pattern.Consume("b,b"));
			ResultAssert.Fails(pattern.Consume("b,a"));
		}
	}
}

File: Tests.FSharp/CaptureTests.fs
namespace Patterns

open System
open Stringier.Patterns
open Defender
open Xunit

type CaptureTests() =
    inherit Trial()

    [<Fact>]
    let ``constructor`` () =
        let mutable capture = ref null
        let _pattern = ('a' || 'b' || 'c') => capture
        ()

    [<Fact>]
    let ``consume`` () =
        let mutable capture = ref null
        
        ('a' || 'b' || 'c') => capture
        |> consumes "a" "a"   
        |> ignore
        capture |> captures "a" |> ignore

        ('a' || 'b' || 'c') => capture >> '!'
        |> consumes "a!" "a!"
        |> fails "a"
        |> ignore
        capture |> captures "a" |> ignore

        ('a' || 'b' || 'c') => capture >> ',' >> capture
        |> consumes "a,a" "a,a"
        |> consumes "b,b" "b,b"
        |> consumes "c,c" "c,c"
        |> fails "b,a"
        |> ignore